src/app/shell/item-comparators.ts

```diff
@@ -240,3 +349,69 @@ export function sortItems(
   );
   return items.toSorted(comparator);
 }
+
+export function groupItems(
+  items: readonly DimItem[],
+  vaultGrouping: string,
+  getTag: (item: DimItem) => TagValue | undefined,
+): readonly VaultGroup[] {
+  const groupingValueGetter = GROUPING_VALUE_GETTERS[vaultGrouping];
+  const iconValueGetter = ICON_VALUE_GETTERS[vaultGrouping];
+  const comparator = GROUP_BY_COMPARATORS[vaultGrouping];
+
+  // If there are no items, or the grouping is not suppored, return all items in a single group
+  if (!items.length || !groupingValueGetter || !comparator) {
+    return [{ groupingValue: undefined, items }];
+  }
+
+  const grouped: MutableVaultGroup[] = [];
+
+  for (const item of items) {
+    const indexOfUngrouped = grouped.findIndex((g) => g.groupingValue === undefined);
+
+    if (!groupingValueGetter) {
+      // If there is not an ungrouped group, create one
```

Comment:
This is an interesting insertion grouping approach, but it might be easier and more efficient to just collect things in a Map?

Reply:
I'm not sure what you mean/how that would work. Can you expand?

Reply:
e.g.:

```ts
const groupedItems = Map.groupBy(items, (i) => groupingValueGetter(item, getTag));
return Array.from(groupedItems.entries(), ([groupingValue, items]) => ({ groupingValue, items, iconValue: iconValueGetter?.(items[0], getTag) })).sort(comparator);
```